agents/ModularAgent.py

```diff
@@ -0,0 +1,649 @@
+	def __init__(self,selection_expression, action, input_rule, selection_var, input_vars, input_args,conditions=[],label=None):
+		self.selection_expression = selection_expression
+		self.action = action
+		self.input_rule = input_rule
+		self.selection_var = selection_var
+		self.input_vars = input_vars
+		self.input_args = input_args
+		self.all_vars = tuple([self.selection_var] + self.input_vars)
+		self.as_tuple = (self.selection_expression,self.action,self.input_rule)
+
+		self.conditions = conditions
+		self.label = label
+		self._how_depth = None
+		self._id_num = None
+
+		self.where = None
+		self.when = None
+		self.which = None
+	
+	# def to_tuple():
+	# 	return (self.selection_expression,self.action,self.input_rule)
+	def to_xml(self,agent=None): #-> needs some way of representing itself including its when/where/how parts
+		pass
+	def get_how_depth(self):
+		if(self._how_depth == None):
+			self._how_depth = compute_exp_depth(self.input_rule)
+		return self._how_depth
+
+	def __hash__(self):
+		return self._id_num
+
+	def __eq__(self,other):
+		return self._id_num == other._id_num
+
+
+
+class Explanation(object):
+	
+	def __init__(self,rhs, mapping):
+		assert isinstance(mapping,dict), "Mapping must be type dict got type %r" % type(mapping)
+		self.rhs = rhs
+		self.mapping = mapping
+		self.selection_literal = mapping[rhs.selection_var]
+		self.input_literals = [mapping[s] for s in rhs.input_vars]#The Literal 
+
+	def compute(self, knowledge_base,operators,epsilon):
+
+		# print("input_rule:", self.skill.input_rule)
+		# print("selection_expression:", self.skill.selection_expression)
+		v = eval_expression([subst(self.mapping,self.rhs.input_rule)], knowledge_base,operators,epsilon)[0]
+
+		return {self.rhs.input_args[0]:v}
+	def conditions_apply(self):
+		return True
+		# for c in self.skill.conditions:
+		# 	c.applies(...)
+	# def to_sai():
+	# 	return SAI(self.input_selections,self.skill.action, self.compute)
+	def to_response(self, knowledge_base,operators,epsilon):
+		# print("SHOULD BE", eval_expression([subst(self.mapping,self.skill.selection_expression)], knowledge_base,operators,epsilon))
+
+		response = {}
+		response['skill_label'] = self.rhs.label
+		response['selection'] = self.selection_literal.replace("?ele-","")
+		# print('selection',response['selection'])
+		response['action'] = self.rhs.action
+		# computed_inputs = self.compute()
+		response['inputs'] = self.compute(knowledge_base,operators,epsilon)#{a: rg_exp[3 + i] for i, a in
+		# print('inputs',response['inputs'])
+							  #enumerate(input_args)}
+		# response['inputs'] = list(rg_exp[3:])
+		response['foci_of_attention'] = []
+		return response
+
+
+	def to_xml(self,agent=None): #-> needs some way of representing itself including its when/where/how parts
+		pass
+
+	def get_how_depth(self):
+		return self.rhs.get_how_depth()
+
+
+	#????? Should this be an object?, Should there be ways of translating between different types of states ???????
```

Comment:
Is this no longer relevant?

Reply:
Having an object for translating between different types of states might still be a reasonable ticket item but I'm not sure there is enough diversity in state formats to warrant it.